### PR TITLE
feat: retrieve participant data

### PR DESCRIPTION
**Pull request by andrewmorris43**
Adds functionality to retrieve the participant and demographic data for use in the cohort distribution service. 

- Uses the existing `IUpdateParticipantData.GetParticipant` method to retrieve the participant.
- Uses the existing `ICreateDemographicData.GetDemographicData` method to retrieve the demographic.
- Uses a new method in `ICreateParticipant` to create a new `CohortDistributionParticipant` which is returned from the function.
- Adds unit tests to cover new functionality.

[DTOSS-3117](https://nhsd-jira.digital.nhs.uk/browse/DTOSS-3117)

## Type of changes

- [ ] Refactoring (non-breaking change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)

## Checklist

- [x] I am familiar with the [contributing guidelines](../docs/CONTRIBUTING.md)
- [x] I have followed the code style of the project
- [x] I have added tests to cover my changes
- [ ] I have updated the documentation accordingly
- [ ] This PR is a result of pair or mob programming

---

## Sensitive Information Declaration

To ensure the utmost confidentiality and protect your and others privacy, we kindly ask you to NOT including [PII (Personal Identifiable Information) / PID (Personal Identifiable Data)](https://digital.nhs.uk/data-and-information/keeping-data-safe-and-benefitting-the-public) or any other sensitive data in this PR (Pull Request) and the codebase changes. We will remove any PR that do contain any sensitive information. We really appreciate your cooperation in this matter.

- [x] I confirm that neither PII/PID nor sensitive data are included in this PR and the codebase changes.

### CODE REVIEW

File: application/CohortManager/src/Functions/screeningDataServices/markParticipantAsIneligible/markParticipantAsIneligible.cs
Comment (andrewmorris43) [inline application/CohortManager/src/Functions/screeningDataServices/markParticipantAsIneligible/markParticipantAsIneligible.cs:1]: The `SuccessfulFunctionRequest` method is no longer used so we can remove it.

File: tests/CohortDistributionTests/CreateCohortDistributionTests/CreateCohortDistributionTests.cs [hunk 1]
```diff
@@ -19,11 +19,16 @@ public class CreateCohortDistributionTests
     private readonly Mock<ICallFunction> _callFunction = new();
     private readonly Mock<ICreateResponse> _createResponse = new();
     private readonly Mock<ILogger<CreateCohortDistribution>> _logger = new();
+
```
Comment (andrewmorris43) [inline tests/CohortDistributionTests/CreateCohortDistributionTests/CreateCohortDistributionTests.cs:22]: Some excess spaces here and just below.

File: application/CohortManager/src/Functions/Shared/Common/CohortDistributionHelper.cs
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Common/CohortDistributionHelper.cs:1]: I don't think these functions should be accepting a full createCohortDistributionFunctionBody object when they only need one or two of the fields from them. seems like its making this helper function very tightly coupled with the cohortDistributionService.

File: tests/CohortDistributionTests/RetrieveParticipantDataTests/RetrieveParticipantDataTests.cs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+{
+    private readonly Mock<ICreateResponse> _createResponse = new();
+    private readonly Mock<ILogger<RetrieveParticipantData>> _logger = new();
+    private readonly RetrieveParticipantData _function;
+    private readonly Mock<FunctionContext> _context = new();
+    private readonly Mock<HttpRequestData> _request;
+    private readonly RetrieveParticipantRequestBody _requestBody;
+    private readonly Mock<IParticipantManagerData> _updateParticipantData = new();
+    private readonly Mock<ICreateDemographicData> _createDemographicData = new();
+    private readonly Mock<ICreateParticipant> _createParticipant = new();
```
Comment (MWClayson-NHS) [inline tests/CohortDistributionTests/RetrieveParticipantDataTests/RetrieveParticipantDataTests.cs:28]: Why is this mocked? There is no external calls in this class.

File: application/CohortManager/src/Functions/Shared/Common/CohortDistributionHelper.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+        var requestBody = JsonSerializer.Serialize(retrieveParticipantRequestBody);
+        var response = await GetResponseAsync(requestBody, Environment.GetEnvironmentVariable("RetrieveParticipantDataURL"));
+        _logger.LogInformation("Called retrieve participant data service");
+
+        if (string.IsNullOrEmpty(response))
+        {
+            throw new Exception("there has been a problem getting participant Data");
+        }
+        _logger.LogInformation("");
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Common/CohortDistributionHelper.cs:36]: why do we have an empty Log statement?

File: application/CohortManager/src/Functions/CohortDistributionServices/CreateCohortDistribution/CreateCohortDistribution.cs [hunk 4]
```diff
@@ -51,10 +53,14 @@ public async Task<HttpResponseData> RunAsync([HttpTrigger(AuthorizationLevel.Ano
 
         try
         {
-            var participantData = await RetrieveParticipantData(requestBody);
-            var serviceProvider = await AllocateServiceProvider(requestBody, participantData);
-            var transformedParticipant = await TransformParticipant(requestBody, serviceProvider, participantData);
+            var participantData = await _CohortDistributionHelper.RetrieveParticipantDataAsync(requestBody);
+
+            var serviceProvider = await _CohortDistributionHelper.AllocateServiceProviderAsync(requestBody, participantData);
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/CohortDistributionServices/CreateCohortDistribution/CreateCohortDistribution.cs:58]: these functions are returning null / empty string when there are non 200 Http Responses. it doesn't seem like they are being handled.

File: application/CohortManager/src/Functions/CohortDistributionServices/CreateCohortDistribution/CreateCohortDistribution.cs [hunk 5]
```diff
@@ -65,105 +71,10 @@ public async Task<HttpResponseData> RunAsync([HttpTrigger(AuthorizationLevel.Ano
...
-        }
-        else
-        {
-            string logMessage = "Transform participant function failed.";
-            _logger.LogError(logMessage);
-            throw new Exception(logMessage);
-        }
-    }
-
     private async Task AddCohortDistribution(CohortDistributionParticipant transformedParticipant)
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/CohortDistributionServices/CreateCohortDistribution/CreateCohortDistribution.cs:74]: Why is this api call left out of the helper file?

Comment (MWClayson-NHS) [reply application/CohortManager/src/Functions/CohortDistributionServices/CreateCohortDistribution/CreateCohortDistribution.cs:74]: also there is no error handling

File: application/CohortManager/src/Functions/Shared/Data/Database/IUpdateParticipantData.cs [hunk 6]
```diff
@@ -1,10 +1,12 @@
 namespace Data.Database;
 
 using Model;
+using NHS.CohortManager.CohortDistribution;
 
-public interface IUpdateParticipantData
+public interface IParticipantManagerData
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Data/Database/IUpdateParticipantData.cs:6]: Aware this wasn't changed by this PR but why are there access modifiers in an interface?

File: application/CohortManager/src/Functions/Shared/Data/Database/ParticipantManagerData.cs
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Data/Database/ParticipantManagerData.cs:1]: @antoinerubeiz is also looking at this file. Might be worth aligning before merging

File: tests/CohortDistributionTests/RetrieveParticipantDataTests/RetrieveParticipantDataTests.cs (hunk 2 above)
Comment (andrewmorris43) [reply tests/CohortDistributionTests/RetrieveParticipantDataTests/RetrieveParticipantDataTests.cs:28]: Maybe I've done this wrong, but the reason I've mocked this is so I can verify the `CreateCohortDistributionParticipantModel` method is called and that the response is as expected.

File: application/CohortManager/src/Functions/Shared/Data/Database/IUpdateParticipantData.cs
Comment (andrewmorris43) [inline application/CohortManager/src/Functions/Shared/Data/Database/IUpdateParticipantData.cs:1]: This file name needs to change to `IParticipantManagerData`.

File: application/CohortManager/src/Functions/Shared/Data/Database/ParticipantManagerData.cs
Comment (SamRobinson75684) [reply application/CohortManager/src/Functions/Shared/Data/Database/ParticipantManagerData.cs:1]: I'll talk to him now

File: application/CohortManager/src/Functions/CohortDistributionServices/CreateCohortDistribution/CreateCohortDistribution.cs [hunk 7]
```diff
@@ -54,13 +54,20 @@ public async Task<HttpResponseData> RunAsync([HttpTrigger(AuthorizationLevel.Ano
...
 
-            var serviceProvider = await _CohortDistributionHelper.AllocateServiceProviderAsync(requestBody, participantData);
+            var serviceProvider = await _CohortDistributionHelper.AllocateServiceProviderAsync(requestBody, participantData.Postcode);
+
+            if (serviceProvider == null)
+            {
+                return _createResponse.CreateHttpResponse(HttpStatusCode.BadRequest, req);
+            }
 
             var transformedParticipant = await _CohortDistributionHelper.TransformParticipantAsync(serviceProvider, participantData);
```
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/CohortDistributionServices/CreateCohortDistribution/CreateCohortDistribution.cs:69]: What happens if transform participant fails or returns null or somthing?

File: application/CohortManager/src/Functions/Shared/Common/CohortDistributionHelper.cs
Comment (MWClayson-NHS) [inline application/CohortManager/src/Functions/Shared/Common/CohortDistributionHelper.cs:1]: Why has all the logging been removed in here?

Comment (SamRobinson75684) [reply application/CohortManager/src/Functions/Shared/Common/CohortDistributionHelper.cs:1]: I did feel as though it was needed and felt it just added extra lines of unwanted code

**Review by MWClayson-NHS — APPROVED** [review]
Looks good to me!

**Pull request merged**